src/webapp/oasis_managers.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+
+import ldap
+
+
+def get_oasis_manager_endpoint_info(global_data, vo, ldappass):
+    """ return list of oasis manager info for endpoint with the structure:
+
+        [ {'ContactID': ContactID, 'Name': Name, 'DNs': DNs,
+           'CILogonID': CILogonID, 'sshPublicKeys': [sshPubKey, ...]}, ... ]
+
+        with one list entry per oasis manager for the given vo, whose
+        contact db info specifies a CILogonID, and whose cilogon info
+        for that CILogonID contains a list of ssh public keys. """
+        
+    managers = get_vo_oasis_managers(global_data, vo)
+    cilogon_id_map = get_cilogon_ldap_id_map(ldappass)
+    ssh_keys_map = cilogon_id_map_to_ssh_keys(cilogon_id_map)
+    contact_cilogon_ids = get_contact_cilogon_id_map(global_data)
+    info = []
+    for manager in managers:
+        if manager['ID'] in contact_cilogon_ids:
+            ContactID = manager['ID']
+            Name = manager['Name']
+            DNs = manager['DNs']
+            CILogonID = contact_cilogon_ids[ContactID]['CILogonID']
+            ssh_keys = ssh_keys_map.get(CILogonID, [])
+            info.append({
+                'ContactID': ContactID,
+                'Name': Name,
+                'DNs': DNs,
+                'CILogonID': CILogonID,
+                'sshPublicKeys': ssh_keys
+            })
+    return info
+
+
+def get_contact_cilogon_id_map(global_data):
+    """ return contacts dict, limited to users with a CILogonID """
+    contacts = global_data.get_contacts_data()
+    return { k: v for k, v in contacts.items() if "CILogonID" in v }
+
+
+def get_vo_oasis_managers(global_data, vo):
+    """return OASIS Managers dict for given vo, if any, else an empty dict"""
+    vos_data = global_data.get_vos_data()
+    if vo in vos_data:
+        if "OASIS" in vos_data[vo]:
+            if "Managers" in vos_data[vo]["OASIS"]:
```

Comment:
The thing you gotta be careful with this is that just because `vos_data[vo]["OASIS"]` exists, it doesn't mean that it's an iterable -- could be a None, for example. Yes, it _should_ be if our data is well-formed, but our data is not necessarily well-formed.

Reply:
Check out `webapp.common.is_null()`, which can check if a dictionary tree exists. If I read it right (sorry for the lack of documentation), `webapp.common.is_null(vos_data, vo, "OASIS", "Managers")` might do what you want.

Reply:
Good tip, i've replaced this with a function to traverse the vos_data dict without causing KeyErrors and with a default if the final result is None.